### PR TITLE
[DependencyInjection] Forbid container cloning

### PR DESCRIPTION
**Pull request by jakzal**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15261 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 1]
```diff
@@ -90,6 +90,11 @@ public function __construct(ParameterBagInterface $parameterBag = null)
         $this->parameterBag = $parameterBag ?: new ParameterBag();
     }
 
+    public function __clone()
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Container.php]: I suggest using a private (final) ``__clone`` method. This way, PHP itself will figure out that the class cannot be cloned (``ReflectionClass::isCloneable`` will return ``false``)

Comment (jakzal) [inline src/Symfony/Component/DependencyInjection/Container.php]: Great suggestion. Applied.

File: src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php [hunk 2]
```diff
@@ -662,6 +662,13 @@ public function testAlias()
         $this->assertTrue($c->has('alias'));
         $this->assertSame($c->get('alias'), $c->get('bar'));
     }
+
+    public function testThatCloningIsNotSupported()
+    {
+        $class = new \ReflectionClass('Symfony\Component\DependencyInjection\Container');
+        $clone = $class->getMethod('__clone');
+        $this->assertTrue($clone->isPrivate());
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: I would assert that ``$class->isCloneable()`` returns false too

Comment (jakzal) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Sure. This check would actually be sufficient. Added.

Comment (jakzal) [inline src/Symfony/Component/DependencyInjection/Tests/ContainerTest.php]: Actually, [isCloneable()](http://php.net/manual/en/reflectionclass.iscloneable.php) is available for PHP >= 5.4.

**Comment by fabpot** [comment]
:+1:

**Comment by jakzal** [comment]
Let's wait for a green build. Previous one failed on PHP 5.3 ;)

**Comment by fabpot** [comment]
What about merging this one on 2.3? Cloning a container does not work anyway, so it should be safe.

**Comment by jakzal** [comment]
I'm fine with merging to 2.3. I just checked and there should be no issues in merging this PR to the 2.3 branch.

We can't get a green build, as tests from the tty group are blocking phpunit process on travis. Once 2.7 is merged to 2.8, there should be no blocking, but tests will still be failing. I'll work on fixing them as soon as 2.8 is updated.

**Comment by fabpot** [comment]
After more thought, I'm going to merge it into 2.8.

**Comment by fabpot** [comment]
Thank you @jakzal.

**Comment by fabpot** [comment]
@jakzal I've just merged 2.7 into 2.8.

**Comment by jakzal** [comment]
@fabpot cool. Looks like the fix is already in 2.8. Tests should be passing now.

**Comment by SongoQ** [comment]
:+1:

**Comment by dosten** [comment]
The `ContainerBuilder` should be clonable?

**Comment by jakzal** [comment]
@dosten we've decided not to support cloning of any containers. See #15261.

**Comment by stof** [comment]
@dosten cloning a ContainerBuilder is a huge pain (see my comment in the issue)

**Pull request merged**